firebaseFunctions/src/index.ts

```diff
@@ -231,3 +197,143 @@ export const createDeal = functions.https.onRequest(
+        return response.sendStatus(403);
+      }
+
+      if (!request.body.address) {
+        return response.sendStatus(400);
+      }
+
+      /**
+       * Address coming from the request body
+       */
+      const address: string = request.body.address;
+
+      // Fail if provided address is not an ethereum address
+      try {
+        getAddress(address);
+      } catch {
+        return response.sendStatus(500);
+      }
+
+      try {
+        /**
+         * Current date as a UTC string, later used as a part of message the user is going to sign
+         */
+        const authenticationRequestDate = new Date().toUTCString();
+
+        // Create/update document for the provided address, storing the current time, to be used to create signature
+        await admin
+          .firestore()
+          .collection(USERS_COLLECTION)
+          .doc(address)
+          .set({authenticationRequestDate});
+
+        return response.status(200).json({ authenticationRequestDate });
+      } catch (error) {
+        functions.logger.error(error);
+        return response.sendStatus(500);
+      }
+    }),
+);
+
+/**
+ * Verifies that a message was signed by the provided address.
+ * Therefore it proofs the ownership of the address.
+ */
+export const verifySignedMessageAndCreateCustomToken = functions.https.onRequest(
+  (request, response) =>
+  // Allow cross-origin requests for this function
+    cors(request, response, async () => {
+      if (request.method !== "POST") {
+        return response.sendStatus(403);
+      }
+
+      if (!request.body.address || !request.body.signature) {
+        return response.sendStatus(400);
+      }
+
+      /**
+       * Address coming from the request body
+       */
+      const address: string = request.body.address;
+
+      // Fail if provided address is not an ethereum address
+      try {
+        getAddress(address);
+      } catch {
+        return response.sendStatus(500);
+      }
+
+      /**
+       * Signature of the message which should be signed by the address, coming from the request body
+       */
+      const signature: string = request.body.signature;
+
+      try {
+        // Get the document with date when authentication process was started for this address
+        const userDocRef = admin.firestore().collection(USERS_COLLECTION).doc(address);
+        const userDoc = await userDocRef.get();
+
+        if (userDoc.exists) {
+          /**
+           * date is part of the message that was signed and it is used to verify the address that signed it
+           */
+          const date = userDoc.data()?.authenticationRequestDate;
+          /**
+           * This is the message user was supposed to sign
+           */
+          const message = `${FIREBASE_MESSAGE_TO_SIGN} ${date}`;
+
+          /**
+           * Recover the address of the account used to create the given signature.
+           */
+          const recoveredAddress = verifyMessage(
```

Comment:
Nit: "recover" somehow suggests, that we lost something

--> `const signerAddress`, and could then remove comment